sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala

```diff
@@ -82,86 +130,169 @@ case class SortMergeJoin(
 
         override final def next(): InternalRow = {
           if (hasNext) {
-            // we are using the buffered right rows and run down left iterator
-            val joinedRow = joinRow(leftElement, rightMatches(rightPosition))
-            rightPosition += 1
-            if (rightPosition >= rightMatches.size) {
-              rightPosition = 0
-              fetchLeft()
-              if (leftElement == null || keyOrdering.compare(leftKey, matchKey) != 0) {
-                stop = false
-                rightMatches = null
+            if (bufferedMatches == null || bufferedMatches.size == 0) {
+              // we just found a row with no join match and we are here to produce a row
+              // with this row and a standard null row from the other side.
+              if (continueStreamed) {
+                val joinedRow = smartJoinRow(streamedElement, bufferedNullRow)
+                fetchStreamed()
+                joinedRow
+              } else {
+                val joinedRow = smartJoinRow(streamedNullRow, bufferedElement)
+                fetchBuffered()
+                joinedRow
+              }
+            } else {
+              // we are using the buffered right rows and run down left iterator
+              val joinedRow = smartJoinRow(streamedElement, bufferedMatches(bufferedPosition))
+              bufferedPosition += 1
+              if (bufferedPosition >= bufferedMatches.size) {
+                bufferedPosition = 0
+                if (joinType != FullOuter || secondStreamedElement == null) {
+                  fetchStreamed()
```

Comment:
I think we should use `boundCondition ` to update `bufferedMatches ` after we `fetchStreamed ()` .Otherwise we may get wrong answer.For example
```sql
table a(key int,value int);table b(key int,value int)
data of a
1 3
1 1
2 1
2 3

data of b
1 1
2 1
select a.key,b.key,a.value-b.value from a left outer join b on a.key=b.key and a.value - b.value > 1
```